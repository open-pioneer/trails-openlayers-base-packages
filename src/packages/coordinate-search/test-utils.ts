// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { act, fireEvent } from "@testing-library/react";

export function showDropdown(projSelect: HTMLElement) {
    // open dropdown to include options in snapshot; react-select creates list of options in dom after opening selection
    act(() => {
        fireEvent.keyDown(projSelect, { key: "ArrowDown" });
    });
}

export function getCurrentOptions() {
    return Array.from(
        document.getElementsByClassName("coordinate-input-select__option")
    ) as HTMLElement[];
}

export function getCurrentOptionValues(options: HTMLElement[]) {
    const values: (string | null)[] = [];
    for (const opt of options) {
        values.push(opt.textContent);
    }
    return values;
}

export function getClearButton(coordinateSearchGroup: Element) {
    const buttonDiv = coordinateSearchGroup.querySelector(".chakra-input__right-element");
    if (!buttonDiv) {
        throw new Error("buttons not rendered");
    }
    const clearButton = buttonDiv.querySelector(".coordinate-input-clear-button");
    if (!clearButton) {
        throw new Error("clear button not rendered");
    }
    return clearButton;
}

export function getCopyButton(coordinateSearchGroup: Element) {
    const buttonDiv = coordinateSearchGroup.querySelector(".chakra-input__right-element");
    if (!buttonDiv) {
        throw new Error("buttons not rendered");
    }
    const copyButton = buttonDiv.querySelector(".coordinate-input-copy-button");
    if (!copyButton) {
        throw new Error("copy button not rendered");
    }
    return copyButton;
}
