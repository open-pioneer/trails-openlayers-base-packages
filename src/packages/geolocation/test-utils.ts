// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import OlMap from "ol/Map";
import { Circle, Fill, Stroke, Style } from "ol/style";
import { vi } from "vitest";
import { GeolocationController } from "./GeolocationController";

export function mockSuccessGeolocation(coords: number[]) {
    vi.spyOn(navigator, "geolocation", "get").mockReturnValue({
        clearWatch() {},
        getCurrentPosition() {},
        watchPosition(success) {
            setTimeout(() => {
                success({
                    coords: {
                        latitude: coords[0],
                        longitude: coords[1],
                        accuracy: 2500,
                        altitude: null,
                        altitudeAccuracy: null,
                        heading: null,
                        speed: null
                    }
                } as GeolocationPosition);
            }, 1);

            return 123;
        }
    } satisfies Partial<Geolocation>);
}

export function mockErrorGeolocation() {
    vi.spyOn(navigator, "geolocation", "get").mockReturnValue({
        clearWatch() {},
        getCurrentPosition() {},
        watchPosition(success, error) {
            setTimeout(() => {
                error?.({
                    code: 2,
                    message: "POSITION_UNAVAILABLE",
                    PERMISSION_DENIED: 1,
                    POSITION_UNAVAILABLE: 2,
                    TIMEOUT: 3
                });
            }, 1);

            return 123;
        }
    } satisfies Partial<Geolocation>);
}

export function setup() {
    const olMap = new OlMap();
    return {
        olMap,
        controller: new GeolocationController(olMap, (error) => {
            console.error("Unexpected error", error);
        })
    };
}

export function setupWithCustomProperties() {
    const olMap = new OlMap();
    const maxZoom: number = getCustomMaxZoom();
    const positionFeatureStyle: Style = getCustomPositionStyle();
    const accuracyFeatureStyle: Style = getCustomAccuracyStyle();
    const trackingOptions: PositionOptions = getCustomTrackingOptions();

    const controller = new GeolocationController(
        olMap,
        (error) => {
            console.error("Unexpected error", error);
        },
        trackingOptions
    );
    controller.setMaxZoom(maxZoom);
    controller.setPositionFeatureStyle(positionFeatureStyle);
    controller.setAccuracyFeatureStyle(accuracyFeatureStyle);
    return {
        olMap,
        controller
    };
}

export function getCustomMaxZoom() {
    return 20;
}
export function getCustomPositionStyle() {
    return new Style({
        image: new Circle({
            fill: new Fill({
                color: "rgba(255,255,255,0.4)"
            }),
            stroke: new Stroke({
                color: "#3399CC",
                width: 1.25
            }),
            radius: 5
        }),
        fill: new Fill({
            color: "rgba(255,255,255,0.4)"
        }),
        stroke: new Stroke({
            color: "#3399CC",
            width: 1.25
        })
    });
}

export function getCustomAccuracyStyle() {
    return new Style({
        image: new Circle({
            fill: new Fill({
                color: "rgba(255,255,255,0.4)"
            }),
            stroke: new Stroke({
                color: "#3399CC",
                width: 1.25
            }),
            radius: 5
        }),
        fill: new Fill({
            color: "rgba(255,255,255,0.4)"
        }),
        stroke: new Stroke({
            color: "#3399CC",
            width: 1.25
        })
    });
}

function getCustomTrackingOptions() {
    return {
        "enableHighAccuracy": true,
        "timeout": 20,
        "maximumAge": 50
    };
}
