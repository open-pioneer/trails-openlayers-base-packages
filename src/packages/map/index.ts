// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
export * from "./api";
