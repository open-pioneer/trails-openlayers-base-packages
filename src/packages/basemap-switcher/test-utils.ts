// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { MapConfig, MapConfigProvider, MapRegistry } from "@open-pioneer/map";
import TileLayer from "ol/layer/Tile";
import OSM from "ol/source/OSM";
import Stamen from "ol/source/Stamen";
import { createService } from "@open-pioneer/test-utils/services";
import { MapRegistryImpl } from "@open-pioneer/map/MapRegistryImpl";
import { PackageContextProviderProps } from "@open-pioneer/test-utils/react";
import { screen, waitFor } from "@testing-library/react";

// used to avoid a "ResizeObserver is not defined" error
import ResizeObserver from "resize-observer-polyfill";
global.ResizeObserver = ResizeObserver;

export interface SimpleMapOptions {
    center?: { x: number; y: number };
    zoom?: number;
}

export const MAP_ID = "test";

export async function waitForMapMount(parentTestId = "base") {
    return await waitFor(async () => {
        const domElement = await screen.findByTestId(parentTestId);
        const container = domElement.querySelector(".ol-viewport");
        if (!container) {
            throw new Error("map not mounted");
        }
        return domElement;
    });
}
export async function setupMap(options?: SimpleMapOptions) {
    const mapId = "test";
    const mapConfig: MapConfig = {
        initialView: {
            kind: "position",
            center: options?.center ?? { x: 847541, y: 6793584 },
            zoom: options?.zoom ?? 10
        },
        projection: "EPSG:3857",
        layers: [
            {
                id: "b-1",
                title: "OSM",
                isBaseLayer: true,
                visible: false,
                layer: new TileLayer({
                    source: new OSM()
                })
            },
            {
                id: "b-2",
                title: "Toner",
                isBaseLayer: true,
                visible: true,
                layer: new TileLayer({
                    source: new Stamen({ layer: "toner" })
                })
            }
        ]
    };

    const registry = await createService(MapRegistryImpl, {
        references: {
            providers: [new MapConfigProviderImpl(mapId, mapConfig)]
        }
    });

    return { mapId, registry };
}

export function createPackageContextProviderProps(
    service: MapRegistry
): PackageContextProviderProps {
    return {
        services: {
            "map.MapRegistry": service
        }
    };
}

class MapConfigProviderImpl implements MapConfigProvider {
    mapId = "default";
    mapConfig: MapConfig;

    constructor(mapId: string, mapConfig?: MapConfig | undefined) {
        this.mapId = mapId;
        this.mapConfig = mapConfig ?? {};
    }

    getMapConfig(): Promise<MapConfig> {
        return Promise.resolve(this.mapConfig);
    }
}
