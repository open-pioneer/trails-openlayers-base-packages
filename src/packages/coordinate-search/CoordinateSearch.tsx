// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { MapModelProps, useMapModel } from "@open-pioneer/map";
import { CommonComponentProps, useCommonComponentProps } from "@open-pioneer/react-utils";
import { FC, useEffect, useState } from "react";
import { Coordinate } from "ol/coordinate";
import { EventsKey } from "ol/events";
import { unByKey } from "ol/Observable";
import OlMap from "ol/Map";
import { CoordinateInput, CoordinatesSelectEvent, ProjectionInput } from "./CoordinateInput";

/**
 * Properties for the {@link CoordinateSearch}.
 */
export interface CoordinateSearchProps extends CommonComponentProps, MapModelProps {
    /**
     * Searchable projections, only projections that are known by the map as projection are shown.
     * Each projection can have an individual precision of coordinates. If no precision is given, the default precision is used.
     */
    projections?: ProjectionInput[];

    /**
     * Optional event that gets called if some coordinates are entered or projection is changed by the user.
     */
    onSelect?: (event: CoordinatesSelectEvent) => void;

    /**
     * Optional event that gets called if the input is cleared.
     */
    onClear?: () => void;
}

/**
 * The `CoordinateSearch` component can be used in an app to search for entered coordinates in a selected projection
 */
export const CoordinateSearch: FC<CoordinateSearchProps> = (props) => {
    const { onSelect, onClear, projections } = props;
    const { containerProps } = useCommonComponentProps("coordinate-search", props);
    const { map } = useMapModel(props);
    const olMap = map?.olMap;

    const { coordinates } = useCoordinates(olMap); // coordinates of the pointer in the map

    return (
        <CoordinateInput
            {...containerProps}
            mapId={props.mapId}
            onSelect={(event: CoordinatesSelectEvent) => {
                if (!map) {
                    return;
                }

                olMap?.getView().setCenter(event.coords);
                onSelect?.(event);
            }}
            onClear={onClear}
            placeholder={coordinates ? coordinates : ""}
            projections={projections}
        />
    );
};

/* get the current coordinates of the pointer on the map */
function useCoordinates(map: OlMap | undefined): { coordinates: Coordinate | undefined } {
    const [coordinates, setCoordinates] = useState<Coordinate | undefined>();

    useEffect(() => {
        if (!map) {
            return;
        }

        const eventsKey: EventsKey = map.on("pointermove", (evt) => {
            setCoordinates(evt.coordinate);
        });

        return () => unByKey(eventsKey);
    }, [map]);

    return { coordinates };
}
