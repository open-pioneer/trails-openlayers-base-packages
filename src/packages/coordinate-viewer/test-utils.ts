// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { OlMapConfigurationProvider } from "@open-pioneer/experimental-ol-map";
import { OlMapRegistry } from "@open-pioneer/experimental-ol-map/services";
import { ServiceOptions } from "@open-pioneer/runtime";
import { PackageContextProviderProps } from "@open-pioneer/test-utils/react";
import { createService } from "@open-pioneer/test-utils/services";
import { waitFor, screen } from "@testing-library/react";
import { MapOptions } from "ol/Map";
import View from "ol/View";
import TileLayer from "ol/layer/Tile";
import OSM from "ol/source/OSM";

// used to avoid a "ResizeObserver is not defined" error
import ResizeObserver from "resize-observer-polyfill";
global.ResizeObserver = ResizeObserver;

export interface SimpleMapOptions {
    center?: [number, number];
    zoom?: number;
}

export const MAP_ID = "test";

export async function waitForMapMount(parentTestId = "base") {
    return await waitFor(async () => {
        const domElement = await screen.findByTestId(parentTestId);
        const container = domElement.querySelector(".ol-viewport");
        if (!container) {
            throw new Error("map not mounted");
        }
        return domElement;
    });
}

export async function setupMap(options?: SimpleMapOptions) {
    const mapOptions: MapOptions = {
        view: new View({
            projection: "EPSG:3857",
            center: options?.center ?? [847541, 6793584],
            zoom: options?.zoom ?? 10
        }),
        layers: [
            new TileLayer({
                source: new OSM(),
                properties: { title: "OSM" }
            })
        ]
    };

    const mapConfigProvider = await createService(MapConfigProvider, {
        properties: {
            mapOptions: mapOptions,
            mapId: MAP_ID
        }
    });
    const registry = await createService(OlMapRegistry, {
        references: {
            providers: [mapConfigProvider]
        }
    });

    return { mapId: MAP_ID, registry };
}

export function createPackageContextProviderProps(
    service: OlMapRegistry
): PackageContextProviderProps {
    return {
        services: {
            "ol-map.MapRegistry": service
        }
    };
}

class MapConfigProvider implements OlMapConfigurationProvider {
    mapId = "default";
    mapOptions: MapOptions = {};

    constructor(options: ServiceOptions) {
        if (options.properties.mapOptions) {
            this.mapOptions = options.properties.mapOptions as MapOptions;
        }
        if (options.properties.mapId) {
            this.mapId = options.properties.mapId as string;
        }
    }

    getMapOptions(): Promise<MapOptions> {
        return Promise.resolve(this.mapOptions);
    }
}
