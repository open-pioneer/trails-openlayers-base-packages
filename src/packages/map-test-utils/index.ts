// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { createService } from "@open-pioneer/test-utils/services";
import { screen, waitFor } from "@testing-library/react";
import TileLayer from "ol/layer/Tile";
import OSM from "ol/source/OSM";
import {
    ExtentConfig,
    InitialViewConfig,
    LayerConfig,
    MapConfig,
    MapConfigProvider,
    MapModel,
    MapRegistry,
    OlMapOptions
} from "@open-pioneer/map";
// Importing internals: needed for test support
import { MapRegistryImpl } from "@open-pioneer/map/services";

export interface SimpleMapOptions {
    center?: { x: number; y: number };
    zoom?: number;
    extent?: ExtentConfig;
    projection?: string;
    layers?: LayerConfig[];
    advanced?: OlMapOptions;

    noInitialView?: boolean;
    noProjection?: boolean;
}

/**
 * Waits until the open layers map has been mounted in the parent with the given id.
 */
export async function waitForMapMount(parentTestId = "base") {
    return await waitFor(async () => {
        const domElement = await screen.findByTestId(parentTestId);
        const container = domElement.querySelector(".ol-viewport");
        if (!container) {
            throw new Error("map not mounted");
        }
        return domElement;
    });
}

/**
 * Waits until the model has an initial extent.
 */
export async function waitForInitialExtent(model: MapModel) {
    if (model.initialExtent) {
        return;
    }

    await new Promise<void>((resolve, reject) => {
        model?.once("changed:initialExtent", () => {
            if (model?.initialExtent) {
                resolve();
            } else {
                reject(new Error("expected a valid extent"));
            }
        });
    });
}

/**
 * Creates a simple map registry service with exactly one map configuration.
 *
 * The map is configured by using the `options` parameter.
 *
 * Returns the map registry and the id of the configured map.
 */
export async function setupMap(options?: SimpleMapOptions) {
    // Always use "test" as mapId for unit tests
    const mapId = "test";

    const getInitialView = (): InitialViewConfig => {
        if (options?.extent) {
            return {
                kind: "extent",
                extent: options.extent
            };
        }
        return {
            kind: "position",
            center: options?.center ?? { x: 847541, y: 6793584 },
            zoom: options?.zoom ?? 10
        };
    };

    const mapConfig: MapConfig = {
        initialView: options?.noInitialView ? undefined : getInitialView(),
        projection: options?.noProjection ? undefined : options?.projection ?? "EPSG:3857",
        layers: options?.layers ?? [
            {
                title: "OSM",
                layer: new TileLayer({
                    source: new OSM()
                })
            }
        ],
        advanced: options?.advanced
    };

    const registry = await createService(MapRegistryImpl, {
        references: {
            providers: [new MapConfigProviderImpl(mapId, mapConfig)]
        }
    });

    return { mapId, registry };
}

/**
 * Creates (service name, service implementation)-pairs suitable for the `services`
 * option of the `PackageContextProvider`.
 *
 * This helper method can be used to avoid hard-coding service names used in the implementation.
 */
export function createServiceOptions(services: { registry: MapRegistry }): Record<string, unknown> {
    return {
        "map.MapRegistry": services.registry
    };
}

class MapConfigProviderImpl implements MapConfigProvider {
    mapId = "default";
    mapConfig: MapConfig;

    constructor(mapId: string, mapConfig?: MapConfig | undefined) {
        this.mapId = mapId;
        this.mapConfig = mapConfig ?? {};
    }

    getMapConfig(): Promise<MapConfig> {
        return Promise.resolve(this.mapConfig);
    }
}
