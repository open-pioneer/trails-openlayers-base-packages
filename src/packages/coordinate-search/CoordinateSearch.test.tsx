// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { MapContainer } from "@open-pioneer/map";
import { createServiceOptions, setupMap, waitForMapMount } from "@open-pioneer/map-test-utils";
import { PackageContextProvider } from "@open-pioneer/test-utils/react";
import { act, render, screen, waitFor } from "@testing-library/react";
import userEvent, { UserEvent } from "@testing-library/user-event";
import { Coordinate } from "ol/coordinate";
import BaseEvent from "ol/events/Event";
import { expect, it, vi } from "vitest";
import { CoordinateSearch } from "./CoordinateSearch";
import {
    getClearButton,
    getCopyButton,
    getCurrentOptions,
    getCurrentOptionValues,
    showDropdown
} from "./test-utils";
import { NumberParser } from "@open-pioneer/core";
import { NumberParserService } from "@open-pioneer/runtime";

it("should successfully create a coordinate search component", async () => {
    const { mapId, injectedServices } = await setUp();

    render(
        <PackageContextProvider services={injectedServices}>
            <CoordinateSearch mapId={mapId} data-testid="coordinate-search" />
        </PackageContextProvider>
    );

    // coordinate search is mounted
    const { coordsSearchDiv } = await waitForCoordinateSearch();
    expect(coordsSearchDiv).toMatchSnapshot();

    // check coordinate search box is available
    expect(coordsSearchDiv.tagName).toBe("DIV");
});

it("should successfully create a coordinate search component with additional css classes", async () => {
    const { mapId, injectedServices } = await setUp();

    render(
        <PackageContextProvider services={injectedServices}>
            <CoordinateSearch mapId={mapId} className="test" data-testid="coordinate-search" />
        </PackageContextProvider>
    );

    const { coordsSearchDiv } = await waitForCoordinateSearch();
    expect(coordsSearchDiv.classList.contains("test")).toBe(true);
    expect(coordsSearchDiv.classList.contains("foo")).toBe(false);
});

it("should successfully create a coordinate search component with projections", async () => {
    const { mapId, injectedServices } = await setUp();

    render(
        <PackageContextProvider services={injectedServices}>
            <CoordinateSearch
                mapId={mapId}
                data-testid="coordinate-search"
                projections={[
                    {
                        label: "EPSG:25832",
                        value: "EPSG:25832"
                    },
                    {
                        label: "WGS 84",
                        value: "EPSG:4326"
                    },
                    {
                        label: "Web Mercator",
                        value: "EPSG:3857"
                    }
                ]}
            />
        </PackageContextProvider>
    );

    const { projSelect } = await waitForCoordinateSearch();
    showDropdown(projSelect);
    const options = getCurrentOptions();
    const values = getCurrentOptionValues(options);

    expect(values).toStrictEqual(["EPSG:25832", "WGS 84", "Web Mercator"]);
});

it("tracks the user's mouse position", async () => {
    const { mapId, injectedServices, registry } = await setUp("de");

    render(
        <PackageContextProvider services={injectedServices} locale="de">
            <MapContainer mapId={mapId} data-testid="map" />
            <CoordinateSearch mapId={mapId} data-testid="coordinate-search" />
        </PackageContextProvider>
    );

    await waitForMapMount("map");
    const { coordInput } = await waitForCoordinateSearch();
    expect(coordInput.getAttribute("placeholder")).toMatchInlineSnapshot('""');

    const map = await registry.expectMapModel(mapId);

    const simulateMove = (x: number, y: number) => {
        const fakeMoveEvent = new BaseEvent("pointermove");
        (fakeMoveEvent as any).coordinate = [x, y];
        map.olMap.dispatchEvent(fakeMoveEvent);
    };

    // Simple move
    act(() => {
        simulateMove(777000, 6698400);
    });
    expect(coordInput.getAttribute("placeholder")).toMatchInlineSnapshot('"6,980 51,434"');

    // Another move + projection change
    act(() => {
        simulateMove(754602, 6664688);
    });
    expect(coordInput.getAttribute("placeholder")).toMatchInlineSnapshot('"6,779 51,245"');
});

it("should display transformed coordinates in selected option", async () => {
    const user = userEvent.setup();
    const { mapId, injectedServices, registry } = await setUp();

    render(
        <PackageContextProvider services={injectedServices}>
            <MapContainer mapId={mapId} data-testid="map" />
            <CoordinateSearch mapId={mapId} data-testid="coordinate-search" />
        </PackageContextProvider>
    );

    await waitForMapMount("map");
    const { coordInput, projSelect } = await waitForCoordinateSearch();
    showDropdown(projSelect);

    const map = await registry.expectMapModel(mapId);

    const simulateMove = (x: number, y: number) => {
        const fakeMoveEvent = new BaseEvent("pointermove");
        (fakeMoveEvent as any).coordinate = [x, y];
        map.olMap.dispatchEvent(fakeMoveEvent);
    };

    let options = getCurrentOptions();

    const option4326 = options.find((option) => option.textContent === "WGS 84");
    if (!option4326) {
        throw new Error("EPSG 4326 missing in options");
    }
    await user.click(option4326);

    // Simple move
    act(() => {
        simulateMove(851594.11, 6789283.95); //map projection is EPSG:3857 (Web Mercator)
    });
    expect(coordInput.getAttribute("placeholder")).toMatchInlineSnapshot('"7.650 51.940"'); //should display EPSG 4326

    showDropdown(projSelect);
    options = getCurrentOptions();
    const option3857 = options.find((option) => option.textContent === "Web Mercator");
    if (!option3857) {
        throw new Error("EPSG 3857 missing in options");
    }
    await user.click(option3857);

    // Simple move
    act(() => {
        simulateMove(851594.11, 6789283.95); //map projection is EPSG:3857 (Web Mercator)
    });
    expect(coordInput.getAttribute("placeholder")).toMatchInlineSnapshot(
        '"851,594.11 6,789,283.95"'
    ); //should display EPSG 3857
});

it(
    "should successfully call onSelect and return coordinates and projection",
    {
        timeout: 10000
    },
    async () => {
        const user = userEvent.setup();
        const { mapId, injectedServices } = await setUp();

        let searchedCoords: Coordinate = [];
        let callbackProj;
        render(
            <PackageContextProvider services={injectedServices}>
                <MapContainer mapId={mapId} data-testid="map" />
                <CoordinateSearch
                    mapId={mapId}
                    data-testid="coordinate-search"
                    onSelect={({ coords, projection }) => {
                        searchedCoords = coords;
                        callbackProj = projection;
                    }}
                />
            </PackageContextProvider>
        );

        await waitForMapMount("map");
        const { coordInput } = await waitForCoordinateSearch();
        await user.type(coordInput, "7 51{enter}");
        expect(searchedCoords).toStrictEqual([779236.4355529151, 6621293.722740165]);
        expect(callbackProj).toBeDefined();
        expect(callbackProj!.getCode()).toBe("EPSG:3857");
    }
);

it(
    "should set the map center to the entered coordinate",
    {
        timeout: 10000
    },
    async () => {
        const user = userEvent.setup();
        const { mapId, injectedServices, registry } = await setUp();
        const map = (await registry.expectMapModel(mapId))?.olMap;

        render(
            <PackageContextProvider services={injectedServices}>
                <MapContainer mapId={mapId} data-testid="map" />
                <CoordinateSearch mapId={mapId} data-testid="coordinate-search" />
            </PackageContextProvider>
        );

        await waitForMapMount("map");
        const { coordInput, coordinateInputGroup } = await waitForCoordinateSearch();

        await input(user, coordInput, "7 51");
        let clearButton = getClearButton(coordinateInputGroup);
        const firstCenter = map.getView().getCenter();
        expect(firstCenter).toEqual([779236.4355529151, 6621293.722740165]);

        await user.click(clearButton);
        await input(user, coordInput, "6 51");
        const secondCenter = map.getView().getCenter();
        expect(secondCenter).toEqual([667916.9447596414, 6621293.722740165]);

        clearButton = getClearButton(coordinateInputGroup);
        await user.click(clearButton);
        await input(user, coordInput, "6b 51"); // wrong input
        const thirdCenter = map.getView().getCenter();
        expect(thirdCenter).toEqual(secondCenter); // map unchanged
    }
);

it("should successfully call onClear if clear button is clicked", async () => {
    const user = userEvent.setup();
    const { mapId, injectedServices } = await setUp();

    let cleared: boolean = false;
    render(
        <PackageContextProvider services={injectedServices}>
            <MapContainer mapId={mapId} data-testid="map" />
            <CoordinateSearch
                mapId={mapId}
                data-testid="coordinate-search"
                onClear={() => {
                    cleared = true;
                }}
            />
        </PackageContextProvider>
    );

    await waitForMapMount("map");
    const { coordInput, coordinateInputGroup } = await waitForCoordinateSearch();

    await user.type(coordInput, "4 5{enter}");
    const clearButton = getClearButton(coordinateInputGroup);
    expect(cleared).toBe(false);

    await user.click(clearButton);
    expect(cleared).toBe(true);
});

it("should successfully copy to clipboard if copy button is clicked", async () => {
    const user = userEvent.setup();
    const { mapId, injectedServices, registry } = await setUp();
    let copiedText = "";

    render(
        <PackageContextProvider services={injectedServices}>
            <MapContainer mapId={mapId} data-testid="map" />
            <CoordinateSearch mapId={mapId} data-testid="coordinate-search" />
        </PackageContextProvider>
    );

    await waitForMapMount("map");
    const { coordinateInputGroup } = await waitForCoordinateSearch();

    const map = await registry.expectMapModel(mapId);

    const simulateMove = (x: number, y: number) => {
        const fakeMoveEvent = new BaseEvent("pointermove");
        (fakeMoveEvent as any).coordinate = [x, y];
        map.olMap.dispatchEvent(fakeMoveEvent);
    };

    // Simple move
    act(() => {
        simulateMove(850000, 6800000); //map projection is EPSG:3857 (Web Mercator)
    });
    vi.spyOn(navigator, "clipboard", "get").mockReturnValue({
        writeText(input: string) {
            copiedText = input;
            return new Promise<void>((resolve) => resolve());
        },
        readText(): Promise<string> {
            return new Promise<string>((resolve) => resolve(""));
        },
        write(): Promise<void> {
            return new Promise<void>((resolve) => resolve());
        },
        addEventListener(): void {},
        dispatchEvent(): boolean {
            return false;
        },
        removeEventListener(): void {},
        read(): Promise<ClipboardItems> {
            return new Promise<ClipboardItems>((resolve) => resolve([]));
        }
    });
    const copyButton = getCopyButton(coordinateInputGroup);
    await user.click(copyButton);
    expect(copiedText).toBe("7.636 51.999");
});

async function waitForCoordinateSearch() {
    const { coordsSearchDiv, coordInput, coordinateInputGroup, projSelect } = await waitFor(
        async () => {
            const coordsSearchDiv = await screen.findByTestId("coordinate-search");

            const coordinateInputGroup = coordsSearchDiv.querySelector(".coordinate-input-group");
            if (!coordinateInputGroup) {
                throw new Error("coordinate input group not rendered");
            }

            const coordInputDiv = coordinateInputGroup.querySelector(
                ".coordinate-input-field-group"
            );
            if (!coordInputDiv) {
                throw new Error("coordinate input field  group not rendered");
            }

            const coordInput = coordInputDiv.querySelector(".chakra-input");
            if (!coordInput) {
                throw new Error("coordinate input field not rendered");
            }

            const projSelect: HTMLElement | null = coordinateInputGroup.querySelector(
                ".coordinate-input-select--has-value"
            );
            if (!projSelect) {
                throw new Error("coordinate input projection select not rendered");
            }

            return { coordsSearchDiv, coordInput, coordinateInputGroup, projSelect };
        }
    );

    return { coordsSearchDiv, coordInput, coordinateInputGroup, projSelect };
}

// A bit faster than typing letters individually with `keyboard`
async function input(user: UserEvent, element: Element, value: string) {
    await user.click(element);
    await user.paste(value);
    await user.keyboard("{Enter}");
}

async function setUp(locale: string = "en") {
    const { mapId, registry } = await setupMap();
    const numberParser = new NumberParser(locale);
    const numberParserService = {
        parseNumber: (number) => {
            return numberParser.parse(number);
        }
    } satisfies Partial<NumberParserService>;

    const injectedServices = {
        ...createServiceOptions({ registry }),
        "runtime.NumberParserService": numberParserService
    };

    return { mapId, injectedServices, registry };
}
