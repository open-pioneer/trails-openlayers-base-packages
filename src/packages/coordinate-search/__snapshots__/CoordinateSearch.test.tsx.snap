// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should successfully create a coordinate search component 1`] = `
<div
  class="coordinate-input coordinate-search css-1xhj18k"
  data-testid="coordinate-search"
>
  <div
    class="chakra-group chakra-tooltip__trigger coordinate-input-group css-r8l1zn"
    data-part="trigger"
    data-scope="tooltip"
    data-state="closed"
    dir="ltr"
    id="tooltip:_r_0_:trigger"
  >
    <div
      class="coordinate-input-select-root chakra-select__root css-11vsvdz"
      data-part="root"
      data-scope="select"
      dir="ltr"
      id="select:_r_1_"
    >
      <div
        class="coordinate-input-select chakra-select__control css-17j1053"
        data-part="control"
        data-scope="select"
        data-state="closed"
        dir="ltr"
        id="select:_r_1_:control"
      >
        <label
          class="chakra-select__label css-164gb0n"
          data-part="label"
          data-scope="select"
          dir="ltr"
          for="select:_r_1_:select"
          id="select:_r_1_:label"
        >
          projections.ariaLabel
        </label>
        <button
          aria-controls="select:_r_1_:content"
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-invalid="false"
          aria-labelledby="select:_r_1_:label"
          class="coordinate-input-select-trigger chakra-select__trigger css-1qq4sc1"
          data-part="trigger"
          data-scope="select"
          data-state="closed"
          dir="ltr"
          id="select:_r_1_:trigger"
          role="combobox"
          type="button"
        >
          <span
            class="coordinate-input-select-value chakra-select__valueText css-176t9sc"
            data-part="value-text"
            data-scope="select"
            dir="ltr"
          >
            WGS 84
          </span>
        </button>
        <div
          class="chakra-select__indicatorGroup css-1fx1ifu"
        >
          <div
            aria-hidden="true"
            class="chakra-select__indicator css-k35okj"
            data-part="indicator"
            data-scope="select"
            data-state="closed"
            dir="ltr"
          >
            <svg
              class="css-s3mb0o"
              viewBox="0 0 24 24"
            >
              <path
                d="m6 9 6 6 6-6"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="chakra-group coordinate-input-field-group css-1w66b0j"
    >
      <div
        class="chakra-field__root css-lesxm"
        data-part="root"
        data-scope="field"
        id="field::_r_2_"
        role="group"
      >
        <input
          aria-label="coordinateInput.ariaLabel"
          autocomplete="off"
          class="chakra-input css-n0ozho"
          data-part="input"
          data-scope="field"
          id="_r_2_"
          placeholder=""
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
